### DISCUSSION TITLE
Returning JSON answer in case of invalid enum

### DISCUSSION THREAD

**Discussion by qarmin** [rwf2/Rocket#2289] [Questions]
- Rocket Version: 0.5.0-rc.2
 ```
#[openapi(tag = "Patient")]
#[post("/patients", data = "<new_patient>")]
pub async fn create_patient(
    db: &State<DatabaseConnection>,
    new_patient: Json<NewPatientInfo>,
) -> Result<Json<NewPatientInfo>, ErrorResponder> {
    Ok(Json(NewPatientInfo))
}
```
when having such code and having such structure
```
#[derive(Serialize, Deserialize, JsonSchema, FromForm)]
pub struct NewPatientInfo {
    pub(crate) sex: Sex,
}
```
with this enum
```
#[derive(FromFormField, JsonSchema, Serialize, Deserialize, Clone, Copy, Debug)]
pub enum Sex {
    MALE,
    FEMALE,
}

impl fmt::Display for Sex {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}
```
in case of giving invalid value to this enum e.g. "MALE2" I have visible error in swagger ui
```
{
  "error": {
    "code": 422,
    "reason": "Unprocessable Entity",
    "description": "The request was well-formed but was unable to be followed due to semantic errors."
  }
}
```
In app logs I see
```
2022-07-24T14:04:05.180374Z  WARN _: Data guard `Json < NewPatientInfo >` failed: Parse("{\n  \"person_id\": \"67445215262\",\n  \"first_name\": \"Roman\",\n  \"last_name\": \"Monitorski\",\n  \"middle_name\": \"Kotarowski\",\n  \"birth_date\": \"2022-02-02\",\n  \"sex\": \"MAFLE\",\n  \"comments\": \"\"\n}", Error("unknown variant `MAFLE`, expected `MALE` or `FEMALE`", line: 7, column: 16)).    
2022-07-24T14:04:05.180413Z  INFO _: Outcome: Failure    
2022-07-24T14:04:05.180434Z  WARN _: No 422 catcher registered. Using Rocket default.    
```

How can I return JSON to swagger ui with this error?

**Comment by edgarogh** [comment]
[`Json`](https://docs.rs/rocket/0.5.0-rc.2/rocket/serde/json/struct.Json.html) if a _form guard_ because [it implements `FromData`](https://docs.rs/rocket/0.5.0-rc.2/rocket/serde/json/struct.Json.html#impl-FromData%3C%27r%3E). _Form guards_ are what you can specify as a `data =` route parameter. As you can see from the implementation, it has an associated error type ; [`rocket::serde::json::Error`](https://docs.rs/rocket/0.5.0-rc.2/rocket/serde/json/enum.Error.html).

Rocket lets you declare a form guard type alone, but in case the guard fails, it will send a 422 Unprocessable entity to the requester by default. You can, however, have your argument be a `Result` (`new_patient: Result<Json<NewPatientInfo>, rocket::serde::json::Error>,` with the correct error type. In this case, you have the freedom to handle parsing errors as you want, and send back a completely custom HTTP response, including a JSON one if you want.
